### PR TITLE
Adding nullable type functionality for HttpRequest Context Type

### PR DESCRIPTION
**Pull request by KellenCarl**
…icAPI.Unshipped.txt to reflect change to nullable type.

- [ ] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [ ] You've included unit or integration tests for your change, where applicable.
- [ ] You've included inline docs for your change, where applicable.
- [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

**PR Title**
Adding nullability annotation on HttpRequest.ContentType 

**PR Description**
Added "?" to make type string nullable for ContentType. Updated PublicAPI.Unshipped.txt to reflect change.  I know there is at least one implementation error, but wanted to let code owners make decisions, provide feedback.
\
Addresses #32097

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=32203) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=32203) before we can accept your contribution.<br/><br/>:x: KellenCarl [sign now](https://cla.dotnetfoundation.org/dotnet/aspnetcore?pullRequest=32203)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/aspnetcore?pullRequest=32203) it.</sub>

File: src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt [hunk 1]
```diff
@@ -18,6 +18,7 @@ Microsoft.AspNetCore.Http.Metadata.IFromServiceMetadata
 Microsoft.AspNetCore.Http.Endpoint.Endpoint(Microsoft.AspNetCore.Http.RequestDelegate? requestDelegate, Microsoft.AspNetCore.Http.EndpointMetadataCollection? metadata, string? displayName) -> void
 Microsoft.AspNetCore.Http.Endpoint.RequestDelegate.get -> Microsoft.AspNetCore.Http.RequestDelegate?
 Microsoft.AspNetCore.Routing.RouteValueDictionary.TryAdd(string! key, object? value) -> bool
+abstract Microsoft.AspNetCore.Http.HttpRequest.ContentType.get -> string?
```
Comment (Tratcher) [inline src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt:21]: Also add a `*REMOVED*` entry above for the old signature.

**Review by Tratcher — COMMENTED** [review]
The build shows two [usages](https://github.com/dotnet/aspnetcore/pull/32203/files) of this property that also need updating, DefaultHttpRequest and HttpLoggingMiddleware.

File: src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (KellenCarl) [reply src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt:22]: *Removed* entry has been added to Unshipped.txt.  I attempted to update the DefaultHttpRequest, and that seemed to go fine.  I am not sure how you want to address the null possibility in the HttpLoggingMiddleware class.

**Comment by BrennanConroy** [comment]
Looks like a couple places need to react to the nullability addition:
src/Middleware/HttpLogging/src/HttpLoggingMiddleware.cs#L118 
src/Hosting/Hosting/src/Internal/HostingRequestStartingLog.cs#L50

**Comment by Tratcher** [comment]
Good, this fix found a bug in the logging middleware (@jkotalik). MediaTypeHelpers.TryGetEncodingForMediaType should allow `string?` and this should be changed to TryParse:
https://github.com/dotnet/aspnetcore/blob/1453c5bd2cf886f6846d2cc00b8106dbd0fdbd53/src/Middleware/HttpLogging/src/MediaTypeHelpers.cs#L31

**Comment by KellenCarl** [comment]
I am currently unable to figure out how to apply the TryParse.  I see the MediaTypeHeaderValue has a method for it, but can't quite get it to work.  I'll go ahead with the other fixes.

File: src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (benaadams) [reply src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt:22]: `ContentType` comes from a `StringValues` so `get` can never return null?

**Comment by KellenCarl** [comment]
I think this is what you were referring to: 
31. MediaTypeHeaderValue.TryParse(contentType, out var mediaType);

When I added that, I received a build error that line 33 would be exposed to the possibility of a null value. And I didn't want to interfere with that logic or implementation.

33.            if (mediaType.Charset.HasValue)

**Comment by KellenCarl** [comment]
This implementation might work, but I am not sure this is what you are going for, or what you want to do with the null value:

if (!MediaTypeHeaderValue.TryParse(contentType, out var mediaType))
            {
                return false;
            }

File: src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (Tratcher) [reply src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt:22]: > `ContentType` comes from a `StringValues` so `get` can never return null?

The default value of a StringValues implicitly converts to null. Right?
https://github.com/dotnet/runtime/blob/41f3d48b520d4cf4f9e051dc07f480dcb9c62e08/src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs#L177-L191

**Comment by Tratcher** [comment]
> This implementation might work, but I am not sure this is what you are going for, or what you want to do with the null value:
> 
> if (!MediaTypeHeaderValue.TryParse(contentType, out var mediaType))
> {
> return false;
> }

That's what I had in mind. It avoids throwing for null, empty, or bad contentTypes.

File: src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt (hunk 1 above)
Comment (benaadams) [reply src/Http/Http.Abstractions/src/PublicAPI.Unshipped.txt:22]: Which is a different value than calling `.ToString()` on it... interesting type 😅 https://github.com/dotnet/runtime/blob/41f3d48b520d4cf4f9e051dc07f480dcb9c62e08/src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs#L168-L171

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks @KellenCarl!

**Comment by KellenCarl** [comment]
Thank you @Tratcher! I appreciate all the help and feedback!

**Pull request merged**